server/src/main/java/org/elasticsearch/gateway/GatewayService.java

```diff
@@ -254,10 +250,15 @@ public void clusterStateProcessed(final String source, final ClusterState oldSta
             logger.info("recovered [{}] indices into cluster_state", newState.metaData().indices().size());
         }
 
+        @Override
+        public void onNoLongerMaster(String source) {
+            logger.debug("stepped down as master before recovering state [{}]", source);
```

Comment:
Why do not we call `resetRecoveredFlags` here?

Reply:
good catch!